src/steam_network/enums.py

```diff
@@ -0,0 +1,179 @@
+            return "steamauthenticator_confirm"
+        elif (self == self.LOGIN):
+            return "login"
+        else:
+            return None
+
+
+    def _add_view(self, args:Dict[str,str]) -> Dict[str, str] :
+        val = self.to_view_string()
+        args["view"] = val if val else "login"
+        return args
+
+    def _get_errored(self, args: Dict[str,str], errored: bool, verbose: bool = False) -> Dict[str, str]:
+        if (errored):
+            args["errored"] = "true"
+        elif (verbose):
+            args["errored"] = "false"
+        return args
+
+    def GetStartUri(self, errored : bool = False, **kwargs:str) -> str:
+        #imho this is the most intuitive way of getting the start url. it's not the most "Pythonic" of means, but it is infinitely more readable.
+
+        #url params go into a dict. urllib.urlencode will autmatically convert a dict to a string of properly concatenated url params ('&'). 
+        #it also escapes any characters that HTTP does not like, which is why we aren't doing it manually. 
+        args : Dict[str, str] = {}
+        #the path to index.html, with a question and a placeholder for all the url parameters. 
+        result : str = str(DIRNAME) + WEBPAGE_RELATIVE_PATH + "?"
+        args = self._add_view(args)
+        args = self._get_errored(args, errored, False)
+
+        for key, value in kwargs.items():
+            if (key not in args):
+                args[key] = value
+
+        #now, convert the dict of url params to a string. replace the placeholder with this string. return the result. 
+        return result + urllib.parse.urlencode(args)
+    
+    def EndUri(self) -> str:
+         if (self == self.TWO_FACTOR_MAIL):
+             return 'two_factor_mail_finished'
+         elif (self == self.TWO_FACTOR_MOBILE):
+             return 'two_factor_mobile_finished'
+         elif (self == self.TWO_FACTOR_CONFIRM):
+             return 'two_factor_confirm_finished'
+         else: #if (self == self.LOGIN):
+             return 'login_finished'
+
+    def GetEndUriRegex(self):
+        if (self != self.TWO_FACTOR_CONFIRM):
+            return ".*" + self.EndUri() + ".*"
+        else:
+            return ".*(" + self.EndUri() + "|" + self.TWO_FACTOR_MAIL.EndUri() + "|" + self.TWO_FACTOR_MOBILE.EndUri() + ").*"
+
+class UserActionRequired(enum.IntEnum):
+    NoActionRequired = 0
+    NoActionConfirmToken = 1
+    NoActionConfirmLogin = 2 #No action required, but we still need to confirm login. New auth workflow requires we poll to get the login info. 
+    TwoFactorRequired = 3 #any form of 2FA required. when set, check the related TwoFactorMethod enum for the thing we need to do. 
+    TwoFactorExpired = 4
+    InvalidAuthData = 5
+
+#We're going to store this in the User Info Cache so we don't need to pass it everywhere. 
+#WARNING! BE VERY CAREFUL WITH THIS: IT APPEARS IN THE USER INFO CACHE! IF IT IS EVER SAVED (toDict method), THIS ENUM BECOMES SOFT IMMUTABLE 
```

Comment:
Just want to point out this comment applies to TwoFactorMethod, i still haven't fully converted all comments to below the class. 

Enums are notoriously annoying when it comes to serialization.  If you rename or reorder an enum, old saved data can point to a different value, which can lead to unexpected behavior. The safest example is if you remove an element from enum, (lets say it had the integer value '5' ) all old saved data that had a value of '5' is now invalid. 

Currently, the TwoFactorMethod is in the user info cache. It's used to tell us which form(s) of 2FA the user has enabled. It's not saved to persistent storage, but it's there. I don't see *why* anyone would save it to storage, but if they do, then change the enum around, it will cause a bunch of broken behavior. It's possible to fix this, of course, because the caches have a "version" so you can quickly convert cached data in "version 1.0" to "version 1.1" or whatever, and if enums changed that's something you could increment the versions for. But i just put the disclaimer in as a CYA in case someone did something later.

Reply:
Well, IntEnum in such usage is only a language alias for numbers, it is quite normal that if you serialize it then you lost representation layer. Steam also have a lot of const values and they don't replace meaning of them; they rather introduce a new one marking the old as deprecated (and there is even don't much value in removing those deprecated values soon).

As for the missing values - adjust the code for your needs. You can do that on serializer/deserializer level or inside the class. There is even dedicated hook in enum for this: https://docs.python.org/3/library/enum.html#enum.Enum._missing_

I didn't understand the initial comment -- I thought it was about updates in the runtime (relating to Python objects mutation). Maybe something like that, less dramatic :)
```suggestion
# Edit with caution because it is deserialized from persistent cache based on value!
```

Reply:
From my dev experience i've seen enum fields get renamed to "depcrecated_" or something similar (as you mention, Steam itself does this in their protos), but the warning was assuming the reader wouldn't know about this practice. i should have explained that better. I'll update the commit accordingly.

Reply:
updated